062-quotas-management.md

```diff
@@ -0,0 +1,163 @@
+   While it can be used by users, it does not follow Strimzi's basic idea of declarative management.
+   Having a dedicated API for quota configuration as part of the `Kafka` CR would make it much easier to configure the default quotas.
+2. The initial releases of the Strimzi Quotas Plugin were very simple.
+   It helps to prevent users producing directly to a broker that was running out of disk space.
+   But it fails to prevent replication of messages produced to other brokers in the Kafka cluster from filling in the disk.
+   It also doesn't support many additional features such as JBOD storage.
+   Because the initial releases were very simple, they had also very simple configuration.
+   And it was possible to configure them in the `.spec.kafka.config` section of the `Kafka` CR.
+   The next - not yet released - version of the Strimzi Quotas plugin brings many improvements.
+   More details about how the new version works can be found in the [Cluster Wide Volume Usage Quota Management](./047-cluster-wide-volume-usage-quota-management.md) proposal.
+   It operates at the level of a whole Kafka cluster and it has improved ability to prevent replicated messages to fill in the broker storage.
+   It also supports JBOD storage.
+   But it has also much more complicated configuration.
+   The plugin is using the Kafka Admin API to gather the information about the disk capacity from the Kafka brokers.
+   And needs therefore to connect to the Kafka cluster as a client.
+   Having the user configure this directly in the `.spec.kafka.config` section would be much more complicated, as the user would need to configure the keystores, truststores, their passwords, bootstrap address, etc.
+   Having a dedicated API for quota configuration would make it much easier to have the operator configure the client automatically.
+
+## Proposal
+
+This proposal suggests to add a new `quotas` API to the `Kafka` custom resource.
+It will be inside the `.spec.kafka` section.
+And it will allow users to configure different types of quotas.
+As many existing Strimzi APIs, it will use the `type` field to differentiate the quota plugin that will be used.
+
+This proposal suggests to initially add support for two Quota plugins.
+1. The built-in Kafka plugin
+2. The Strimzi Quotas plugin
+
+The first supported plugin - Kafka's built-in plugin - will be configured as `type: kafka`.
+The API will allow to configure the default user quotas for this plugin in the same format as the `KafkaUser` API allows it today:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    quotas:
+      type: kafka
+      producerByteRate: 1048576
+      consumerByteRate: 2097152
+      requestPercentage: 55
+      controllerMutationRate: 50
+    # ... 
+  # ...
+```
+
+Based on this configuration, the operator will:
+* Configure the default quota plugin (make sure the `client.quota.callback.class` option is not set)
```

Comment:
will `client.quota.callback.class` be forbidden or still usable but the operator checks that both `quotas` and `client.quota.callback.class` are not set?

Reply:
No. That would make this backward incompatible. It should be implicit from the backward compatibility part of the proposal. But I added it explicitly as well.